Mathlib/Order/LiminfLimsup.lean

```diff
@@ -0,0 +1,1242 @@
+#align filter.mono_bliminf' Filter.mono_bliminf'
+
+theorem mono_bliminf (h : ∀ x, p x → u x ≤ v x) : bliminf u f p ≤ bliminf v f p :=
+  mono_bliminf' <| eventually_of_forall h
+#align filter.mono_bliminf Filter.mono_bliminf
+
+theorem bliminf_antitone_filter (h : f ≤ g) : bliminf u g p ≤ bliminf u f p :=
+  supₛ_le_supₛ fun _ ha => ha.filter_mono h
+#align filter.bliminf_antitone_filter Filter.bliminf_antitone_filter
+
+theorem blimsup_monotone_filter (h : f ≤ g) : blimsup u f p ≤ blimsup u g p :=
+  infₛ_le_infₛ fun _ ha => ha.filter_mono h
+#align filter.blimsup_monotone_filter Filter.blimsup_monotone_filter
+
+
+-- @[simp] -- Porting note: simp_nf linter, lhs simplifies, added _aux version below
+theorem blimsup_and_le_inf : (blimsup u f fun x => p x ∧ q x) ≤ blimsup u f p ⊓ blimsup u f q :=
+  le_inf (blimsup_mono <| by tauto) (blimsup_mono <| by tauto)
+#align filter.blimsup_and_le_inf Filter.blimsup_and_le_inf
+
+@[simp]
+theorem bliminf_sup_le_inf_aux :
+  (blimsup u f fun x => p x ∧ q x) ≤ blimsup u f p
+    ∧ (blimsup u f fun x => p x ∧ q x) ≤ blimsup u f q := by
+  rw [←le_inf_iff]
+  exact blimsup_and_le_inf
+
+-- @[simp] -- Porting note: simp_nf linter, lhs simplifies, added _aux simp version below
+theorem bliminf_sup_le_and : bliminf u f p ⊔ bliminf u f q ≤ bliminf u f fun x => p x ∧ q x :=
+  blimsup_and_le_inf (α := αᵒᵈ)
+#align filter.bliminf_sup_le_and Filter.bliminf_sup_le_and
+
+@[simp]
+theorem bliminf_sup_le_and_aux :
+  (bliminf u f p ≤ bliminf u f fun x => p x ∧ q x)
+    ∧ bliminf u f q ≤ bliminf u f fun x => p x ∧ q x := by
+  rw [←sup_le_iff]
+  exact bliminf_sup_le_and
+
+/-- See also `Filter.blimsup_or_eq_sup`. -/
+-- @[simp] -- Porting note: simp_nf linter, lhs simplifies, added _aux simp version below
+theorem blimsup_sup_le_or : blimsup u f p ⊔ blimsup u f q ≤ blimsup u f fun x => p x ∨ q x :=
+  sup_le (blimsup_mono <| by tauto) (blimsup_mono <| by tauto)
+#align filter.blimsup_sup_le_or Filter.blimsup_sup_le_or
+
+@[simp]
+theorem blimsup_sup_le_or_aux :
+  (blimsup u f p ≤ blimsup u f fun x => p x ∨ q x)
+    ∧ blimsup u f q ≤ blimsup u f fun x => p x ∨ q x := by
+  rw [←sup_le_iff]
+  exact blimsup_sup_le_or
+
+/-- See also `Filter.bliminf_or_eq_inf`. -/
+-- @[simp] -- Porting note: simp_nf linter, lhs simplifies, added _aux simp version below
+theorem bliminf_or_le_inf : (bliminf u f fun x => p x ∨ q x) ≤ bliminf u f p ⊓ bliminf u f q :=
+  blimsup_sup_le_or (α := αᵒᵈ)
+#align filter.bliminf_or_le_inf Filter.bliminf_or_le_inf
+
+@[simp]
+theorem bliminf_or_le_inf_aux :
+  (bliminf u f fun x => p x ∨ q x) ≤ bliminf u f p
+    ∧ (bliminf u f fun x => p x ∨ q x) ≤ bliminf u f q := by
+  rw [←le_inf_iff]
+  exact bliminf_or_le_inf
+
+/- Porting note: Replaced `e` with `FunLike.coe e` to override the strange
+ coercion to `↑(RelIso.toRelEmbedding e).toEmbedding`.-/
+theorem OrderIso.apply_blimsup [CompleteLattice γ] (e : α ≃o γ) :
+    FunLike.coe e (blimsup u f p) = blimsup ((FunLike.coe e) ∘ u) f p := by
```

Comment:
Does it work to write `(e : α -> γ)` instead of the `FunLike.coe`? That's a bit more readable.

Reply:
Thanks for all the very good comments! This one is the only one that doesn't work, `(e : α -> γ)` is automatically coerced to the unpleasant `↑(RelIso.toRelEmbedding e).toEmbedding` expression. This problem is also discussed in [this Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/CoeFun), and due to an undesirable CoeFun instance.